overlord/devicestate/handlers_systems.go

```diff
@@ -145,8 +145,44 @@ func (m *DeviceManager) doCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) (err
 
 	// get all infos
 	infoGetter := func(name string) (info *snap.Info, present bool, err error) {
-		// snap may be present in the system in which case info comes
-		// from snapstate
+		// snaps are either being fetched or present in the system
+
+		if isRemodel {
+			// in a remodel scenario, the snaps may need to be
+			// fetched and thus their content can be different from
+			// what we have in already installed snaps, so we should
+			// first check the download tasks before consulting
+			// snapstate
+			logger.Debugf("requested info for snap %q being installed during remodel", name)
+			for _, tskID := range setup.SnapSetupTasks {
+				taskWithSnapSetup := st.Task(tskID)
+				snapsup, err := snapstate.TaskSnapSetup(taskWithSnapSetup)
```

Comment:
The switch-snap-channel task is produced by the update code upon consulting the store first, so the local revision must be the same as the one in the target channel. In this case there would be no associated download task for this snap, thus the code will eventually take the locally installed snap file for the recovery system.

Reply:
you mean that for that case it doesn't matter if we run this code or the code below? the log message might or might not be confusing then though?

Reply:
I'll extend the logging.

As for tasks, I should probably document that in more detail. When we have a switch-snap-channel, then there is no download checks done edge, then the task ID of switch-snap-channel will not be recorded as a task carrying snapsup for the recovery task. So even though switch-snap-channel actually has snapsup, it will not be looked at when iterating this loop because there is no real download taking place and we take it from the snapstate.CurrentInfo().

In #11085 we actually introduce the edge, and thus will hit the code path here, but snapsup already contains all the relevant data (name & revision) which allows us to identify the mount file of the snap.

Reply:
so we concluded this is fine as is but #11085 needs more work